libraries/USBSerial/src/usbd_cdc_if.c

```diff
@@ -302,137 +316,130 @@ uint8_t CDC_Transmit_FS(uint8_t* Buf, uint16_t Len)
 }
 
-void CDC_flush(void)
-{
+void CDC_flush(void) {
   uint8_t status;
 
-  if(UserTxBufPtrOut != UserTxBufPtrIn)
-  {
-    if(UserTxBufPtrOut > UserTxBufPtrIn) /* Roll-back */
+  if (UserTxBufPtrOut != UserTxBufPtrIn) {
+    if (UserTxBufPtrOut > UserTxBufPtrIn) /* Roll-back */
     {
-      memcpy((uint8_t*)&StackTxBufferFS[0], (uint8_t*)&UserTxBufferFS[UserTxBufPtrOut], (APP_TX_DATA_SIZE - UserTxBufPtrOut));
-
-      memcpy((uint8_t*)&StackTxBufferFS[APP_TX_DATA_SIZE - UserTxBufPtrOut], (uint8_t*)&UserTxBufferFS[0], UserTxBufPtrIn);
-
-      USBD_CDC_SetTxBuffer(&hUSBD_Device_CDC, (uint8_t*)&StackTxBufferFS[0], (APP_TX_DATA_SIZE - UserTxBufPtrOut + UserTxBufPtrIn));
-
-      do {
-        status = USBD_CDC_TransmitPacket(&hUSBD_Device_CDC);
-      } while(status == USBD_BUSY);
-
-      if(status == USBD_OK)
-      {
-        UserTxBufPtrOut = UserTxBufPtrIn;
-      }
+      memcpy((uint8_t *)&StackTxBufferFS[0],
+             (uint8_t *)&UserTxBufferFS[UserTxBufPtrOut],
+             (APP_TX_DATA_SIZE - UserTxBufPtrOut));
+      memcpy((uint8_t *)&StackTxBufferFS[APP_TX_DATA_SIZE - UserTxBufPtrOut],
+             (uint8_t *)&UserTxBufferFS[0], UserTxBufPtrIn);
+
+      USBD_CDC_SetTxBuffer(
+          &hUSBD_Device_CDC, (uint8_t *)&StackTxBufferFS[0],
+          (APP_TX_DATA_SIZE - UserTxBufPtrOut + UserTxBufPtrIn));
+    } else {
+      USBD_CDC_SetTxBuffer(&hUSBD_Device_CDC,
+                           (uint8_t *)&UserTxBufferFS[UserTxBufPtrOut],
+                           (UserTxBufPtrIn - UserTxBufPtrOut));
     }
-    else
-    {
-      USBD_CDC_SetTxBuffer(&hUSBD_Device_CDC, (uint8_t*)&UserTxBufferFS[UserTxBufPtrOut], (UserTxBufPtrIn - UserTxBufPtrOut));
 
-	  do {
+    do {
+      if (lineState == 0) // Device disconnected
+        status = USBD_OK;
+      else
         status = USBD_CDC_TransmitPacket(&hUSBD_Device_CDC);
-      } while(status == USBD_BUSY);
+    } while (status == USBD_BUSY);
 
-      if(status == USBD_OK)
-      {
-        UserTxBufPtrOut = UserTxBufPtrIn;
-      }
+    if (status == USBD_OK) {
+      UserTxBufPtrOut = UserTxBufPtrIn;
     }
   }
 }
 
-void CDC_disable_TIM_Interrupt(void)
-{
-  HAL_NVIC_DisableIRQ(TIM6_DAC_IRQn);
-}
-
-void CDC_enable_TIM_Interrupt(void)
-{
-  HAL_NVIC_EnableIRQ(TIM6_DAC_IRQn);
+void CDC_disable_TIM_Interrupt(void) { HAL_NVIC_DisableIRQ(TIM6_DAC_IRQn); }
+
+void CDC_enable_TIM_Interrupt(void) { HAL_NVIC_EnableIRQ(TIM6_DAC_IRQn); }
+
+void CDC_resume_receive(void) {
+  if (receiveSuspended) {
+    if ((UserRxBufPtrOut + APP_RX_DATA_SIZE - UserRxBufPtrIn - 1) %
+                APP_RX_DATA_SIZE +
+            1 >=
+        USB_OTG_FS_MAX_PACKET_SIZE) {
+      USBD_CDC_ReceivePacket(
+          &hUSBD_Device_CDC); // Initiate next USB packet transfer once a packet
+                              // is received and there is enouch space in the
+                              // buffer
+      receiveSuspended = false;
+      HAL_GPIO_WritePin(GPIOD, GPIO_PIN_10, GPIO_PIN_RESET);
```

Comment:
@hasenbanck ? @ktand ? Why acting on PD10?

Reply:
This change came @ktand 's improvement branch I ported. Maybe it's a change for his particular implementation? Oddly enough it worked on my end without modification.